src/main/scala/com/gradle/internal/Utils.scala

```diff
@@ -4,47 +4,56 @@ import java.io.{FileInputStream, IOException, UnsupportedEncodingException}
 
 object Utils {
 
-  private val GIT_REPO_URI_REGEX = "^(?:(?:https://|git://)|(?:ssh)?.*?@)(.*?(?:github|gitlab).*?)(?:/|:[0-9]*?/|:)(.*?)(?:\\.git)?$".r
-
-  private[gradle] def sysPropertyOrEnvVariable(sysPropertyName: String): Option[String] = sysPropertyOrEnvVariable(sysPropertyName, toEnvVarName(sysPropertyName))
-
-  private[gradle] def booleanSysPropertyOrEnvVariable(sysPropertyName: String): Option[Boolean] = booleanSysPropertyOrEnvVariable(sysPropertyName, toEnvVarName(sysPropertyName))
+  object Env {
+    case class Key[T](name: String) {
+      def envVarName: String = toEnvVarName(name)
+    }
 
-  private def toEnvVarName(sysPropertyName: String) = sysPropertyName.toUpperCase.replace('.', '_')
+    object SystemEnvironment extends Env {
+      override def fromEnvironment(name: String): Option[String] = sys.env.get(name)
+      override def fromProperties(name: String): Option[String] = sys.props.get(name)
+    }
 
-  private[gradle] def sysPropertyOrEnvVariable(sysPropertyName: String, envVarName: String): Option[String] = {
-    sysProperty(sysPropertyName).orElse(envVariable(envVarName))
+    private def toEnvVarName(name: String): String = name.toUpperCase.replace('.', '_')
   }
 
-  private[gradle] def booleanSysPropertyOrEnvVariable(sysPropertyName: String, envVarName: String): Option[Boolean] = {
-    booleanSysProperty(sysPropertyName).orElse(booleanEnvVariable(envVarName))
-  }
+  abstract class Env {
+    protected def fromEnvironment(name: String): Option[String]
+    protected def fromProperties(name: String): Option[String]
 
-  private[gradle] def envVariable(name: String): Option[String] = sys.env.get(name)
+    final def envVariable[T](key: Env.Key[T])(implicit parse: FromString[T]): Option[T] =
+      fromEnvironment(key.envVarName).flatMap(parse(_))
+    final def envVariable[T](name: String)(implicit parse: FromString[T]): Option[T] =
+      fromEnvironment(name).flatMap(parse(_))
 
-  private[gradle] def booleanEnvVariable(name: String): Option[Boolean] = envVariable(name).map(parseBoolean)
+    final def sysProperty[T](key: Env.Key[T])(implicit parse: FromString[T]): Option[T] =
+      fromProperties(key.name).flatMap(parse(_))
+    final def sysProperty[T](name: String)(implicit parse: FromString[T]): Option[T] =
+      fromProperties(name).flatMap(parse(_))
 
-  private[gradle] def sysProperty(name: String): Option[String] = sys.props.get(name)
+    final def sysPropertyOrEnvVariable[T: FromString](name: String): Option[T] = sysPropertyOrEnvVariable(
+      Env.Key[T](name)
+    )
+    final def sysPropertyOrEnvVariable[T: FromString](key: Env.Key[T]): Option[T] =
+      sysProperty(key).orElse(envVariable(key))
 
-  private[gradle] def booleanSysProperty(name: String): Option[Boolean] = sysProperty(name).map(parseBoolean)
+  }
 
-  private[gradle] def isNotEmpty(value: Option[String]): Boolean = value.exists(_.nonEmpty)
+  private val GIT_REPO_URI_REGEX =
```

Comment:
can we move this to the top of the page in order to keep consistency with the Java versions as much as possible?